boot/seal.go

```diff
@@ -25,65 +25,216 @@ import (
+	}
+
+	recoveryBootChain, err := tbl.RecoveryBootChain("")
+	if err != nil {
+		return nil, err
+	}
+
+	// get asset chains
+	assetChain, kbf, err := buildBootAssets(recoveryBootChain, modeenv)
+	if err != nil {
+		return nil, err
+	}
+
+	return &bootChain{
+		BrandID:        model.BrandID(),
+		Model:          model.Model(),
+		Grade:          string(model.Grade()),
+		ModelSignKeyID: model.SignKeyID(),
+		AssetChain:     assetChain,
+		Kernel:         seedKernel.Path,
+		KernelRevision: kernelRev,
+		KernelCmdlines: []string{cmdline},
+		model:          model,
+		kernelBootFile: bootloader.NewBootFile(seedKernel.Path, kbf.Path, kbf.Role),
+	}, nil
+}
+
+func buildRunModeBootChains(rbl, bl bootloader.Bootloader, model *asserts.Model, modeenv *Modeenv, cmdline string) ([]bootChain, error) {
+	tbl, ok := rbl.(bootloader.TrustedAssetsBootloader)
+	if !ok {
+		return nil, fmt.Errorf("recovery bootloader doesn't support trusted assets")
+	}
+	runModeBootChain, err := tbl.BootChain(bl, "")
+	if err != nil {
+		return nil, err
+	}
+
+	// get asset chains
+	assetChain, kbf, err := buildBootAssets(runModeBootChain, modeenv)
+	if err != nil {
+		return nil, err
+	}
+
+	chains := make([]bootChain, 0, len(modeenv.CurrentKernels))
+	for _, k := range modeenv.CurrentKernels {
+		info, err := snap.ParsePlaceInfoFromSnapFileName(k)
+		if err != nil {
+			return nil, err
+		}
+		var kernelRev string
+		if info.SnapRevision().Store() {
+			kernelRev = info.SnapRevision().String()
+		}
+		chains = append(chains, bootChain{
+			BrandID:        model.BrandID(),
+			Model:          model.Model(),
+			Grade:          string(model.Grade()),
+			ModelSignKeyID: model.SignKeyID(),
+			AssetChain:     assetChain,
+			Kernel:         k,
+			KernelRevision: kernelRev,
+			KernelCmdlines: []string{cmdline},
+			model:          model,
+			kernelBootFile: bootloader.NewBootFile(info.MountFile(), kbf.Path, kbf.Role),
+		})
+	}
+	return chains, nil
+}
+
+func buildBootAssets(bootFiles []bootloader.BootFile, modeenv *Modeenv) (assets []bootAsset, kernel bootloader.BootFile, err error) {
+	assets = make([]bootAsset, len(bootFiles)-1)
+
+	for i, bf := range bootFiles[:len(bootFiles)-1] {
+		name := filepath.Base(bf.Path)
+		var hashes []string
+		var ok bool
+		if bf.Role == bootloader.RoleRecovery {
+			hashes, ok = modeenv.CurrentTrustedRecoveryBootAssets[name]
+		} else {
+			hashes, ok = modeenv.CurrentTrustedBootAssets[name]
+		}
+		if !ok {
+			return nil, kernel, fmt.Errorf("cannot find asset %s in modeenv", name)
+		}
+		assets[i] = bootAsset{
+			Role:   bf.Role,
+			Name:   name,
+			Hashes: hashes,
+		}
+	}
+
+	return assets, bootFiles[len(bootFiles)-1], nil
+}
+
+func sealKeyParams(pbc predictableBootChains, roleToBlName map[bootloader.Role]string) (*secboot.SealKeyParams, error) {
+	modelParams := make([]*secboot.SealKeyModelParams, 0, len(pbc))
+	for _, bc := range pbc {
+		loadChains, err := bootAssetsToLoadChains(bc.AssetChain, bc.kernelBootFile, roleToBlName)
```

Comment:
not a blocker but we should try to make one single SealKeyModelParams for the same model, given that pbc is order by model and also we have the model assetions it shouldn't be too hard

Reply:
Added a TODO